src/util.rs

```diff
@@ -107,6 +128,103 @@ impl Default for Digest {
     }
 }
 
+#[cfg(unix)]
+mod file_time_stamp {
+    use std::path::Path;
+
+    #[derive(Clone, PartialEq, Eq, Debug)]
+    pub struct FileTimeStamp {
+        dev: u64,
+        ino: u64,
+        size: u64,
+        mtime: i64,
+        mtime_nsec: i64,
+    }
+
+    impl FileTimeStamp {
+        pub fn from_path(path: &Path) -> std::io::Result<FileTimeStamp> {
+            use std::os::unix::fs::MetadataExt;
+
+            let metadata = std::fs::metadata(path)?;
+
+            Ok(FileTimeStamp {
+                dev: metadata.dev(),
+                ino: metadata.ino(),
+                size: metadata.size(),
+                mtime: metadata.mtime(),
+                mtime_nsec: metadata.mtime_nsec(),
+            })
+        }
+    }
+}
+
+#[cfg(not(unix))]
+mod file_time_stamp {
+    use filetime::FileTime;
+    use std::path::Path;
+
+    #[derive(Clone, PartialEq, Eq, Debug)]
+    pub struct FileTimeStamp(FileTime);
+
+    impl FileTimeStamp {
+        pub fn from_path(path: &Path) -> std::io::Result<FileTimeStamp> {
+            let metadata = std::fs::metadata(path)?;
+            Ok(FileTimeStamp(FileTime::from_last_modification_time(
+                &metadata,
+            )))
+        }
+    }
+}
+
+type FileTimeStamp = file_time_stamp::FileTimeStamp;
+
+#[derive(Clone)]
+struct CachedFileHash {
+    timestamp: FileTimeStamp,
+    hash: Arc<String>,
+}
+
+#[derive(Default)]
+struct FileHashCache {
+    hashes: Arc<Mutex<HashMap<PathBuf, CachedFileHash>>>,
```

Comment:
Without looking too deeply into this, I feel like this shouldn't be free to grow indefinitely, and should be a LRU cache of some sort.